lucille-core/src/main/java/com/kmwllc/lucille/stage/BuildNested.java

```diff
@@ -0,0 +1,330 @@
+      }
+
+      String[] parts = trimmedDest.split("\\.");
+      for (String p : parts) {
+        if (p.isEmpty()) {
+          throw new StageException("Invalid destination '" + trimmedDest + "' (empty segment).");
+        }
+      }
+
+      List<String> key = List.of(parts);
+      if (out.containsKey(key)) {
+        throw new StageException("Duplicate destination '" + dest + "'.");
+      }
+
+      out.put(key, trimmedSrc);
+    }
+
+    return out;
+  }
+
+  @Override
+  public Iterator<Document> processDocument(Document doc) throws StageException {
+    final int n = pickNumObjects();
+    ArrayNode arr = MAPPER.createArrayNode();
+
+    // Iterate over each nested object
+    for (int i = 0; i < n; i++) {
+      ObjectNode entity = MAPPER.createObjectNode();
+      boolean wroteAny = false;
+
+      // Iterate over each entry for the object
+      for (Map.Entry<List<String>, String> e : parsedEntries.entrySet()) {
+        List<String> keyParts = e.getKey();
+        String sourceField = e.getValue();
+
+        // Assign if generators contains the source field name
+        String genKey = (!isBlank(sourceField) && generators.containsKey(sourceField)) ? sourceField : null;
+
+        // Get value from source field if present, otherwise generator
+        JsonNode valNode = null;
+        boolean hasSource = !isBlank(sourceField) && doc.has(sourceField);
+        if (hasSource) {
+          valNode = doc.getJson(sourceField);
+        } else if (genKey != null) {
+          valNode = generateWith(genKey);
+        }
+
+        if (!hasSource && genKey == null) {
+          String destPathStr = String.join(".", keyParts);
+          throw new StageException("Missing value for '" + destPathStr +
+              "' (source='" + sourceField + "') and no generator available.");
+        }
+
+        if (valNode == null || valNode.isNull()) {
+          log.warn("Value for '{}' resolved to null ({}).", sourceField, doc.getId());
+          continue;
+        }
+
+        // Write value at the destination
+        setNested(entity, keyParts, valNode);
+        wroteAny = true;
+      }
+
+      if (wroteAny) {
+        arr.add(entity);
+      }
+    }
+
+    doc.setField(targetField, arr);
+    return null;
+  }
+
+  // Pick the number of objects
+  private int pickNumObjects() {
+    if (numObjects != null) {
+      return numObjects;
+    }
+
+    if (minNumObjects != null) {
+      return ThreadLocalRandom.current().nextInt(minNumObjects, maxNumObjects + 1);
+    }
+
+    return 1;
+  }
+
+  // Generate a value with the provided stage
+  private JsonNode generateWith(String genKey) throws StageException {
+    Stage gen = generators.get(genKey);
+
+    if (gen == null) {
+      return null;
+    }
+
+    // Run the generator on the current doc once
+    gen.processDocument(genDoc);
+    JsonNode value = genDoc.getJson(GEN_OUT_FIELD);
+
+    // Clean up temp field
+    if (genDoc.has(GEN_OUT_FIELD)) {
```

Comment:
might be safer to do this before calling gen.processDocument instead of after
that way if the generator has append instead of overrwrite behavior, we can be sure the GEN_OUT_FIELD wont be there